dont sent kicks upstream for local channels

diff -r b3e677e85b26 ircd/gline.c
--- a/ircd/gline.c
+++ b/ircd/gline.c
@@ -234,7 +234,11 @@
         nmember=member->next_member;
         if (!MyUser(member->user) || IsZombie(member) || IsAnOper(member->user))
           continue;
-        sendcmdto_serv_butone(&me, CMD_KICK, NULL, "%H %C :Badchanneled (%s)", chptr, member->user, gline->gl_reason);
+
+        /* do not send kicks upstream for local channels */
+        if (!IsLocalChannel(chptr->chname))
+          sendcmdto_serv_butone(&me, CMD_KICK, NULL, "%H %C :Badchanneled (%s)", chptr, member->user, gline->gl_reason);
+
         sendcmdto_channel_butserv_butone(&me, CMD_KICK, chptr, NULL, 0, "%H %C :Badchanneled (%s)", chptr, member->user, gline->gl_reason);
         make_zombie(member, member->user, &me, &me, chptr);
         retval=1;
