# HG changeset patch
# Parent 6ddd72a8d4f99bf688363698f44a1c5d81bbcdf9

diff -r 6ddd72a8d4f9 ircd/m_die.c
--- a/ircd/m_die.c	Sun Jul 14 17:52:57 2013 +0100
+++ b/ircd/m_die.c	Sun Jul 14 17:53:48 2013 +0100
@@ -106,6 +106,12 @@
   if (!HasPriv(sptr, PRIV_DIE))
     return send_reply(sptr, ERR_NOPRIVILEGES);
 
+  if (parc < 2 || ircd_strcmp(parv[1],cli_name(&me))) {
+    sendcmdto_one(&me, CMD_NOTICE, sptr, "%C :This will terminate the server %s", sptr, cli_name(&me));
+    sendcmdto_one(&me, CMD_NOTICE, sptr, "%C :If you are sure this is what you want, use /quote DIE %s", sptr, cli_name(&me));
+    return 0;
+  }
+
   for (i = 0; i <= HighestFd; i++)
   {
     if (!(acptr = LocalClientArray[i]))
diff -r 6ddd72a8d4f9 ircd/m_restart.c
--- a/ircd/m_restart.c	Sun Jul 14 17:52:57 2013 +0100
+++ b/ircd/m_restart.c	Sun Jul 14 17:53:48 2013 +0100
@@ -86,6 +86,7 @@
 #include "ircd_log.h"
 #include "ircd_reply.h"
 #include "ircd_string.h"
+#include "msg.h"
 #include "numeric.h"
 #include "numnicks.h"
 #include "send.h"
@@ -100,6 +101,12 @@
   if (!HasPriv(sptr, PRIV_RESTART))
     return send_reply(sptr, ERR_NOPRIVILEGES);
 
+  if (parc < 2 || ircd_strcmp(parv[1],cli_name(&me))) {
+    sendcmdto_one(&me, CMD_NOTICE, sptr, "%C :This will restart the server %s", sptr, cli_name(&me));
+    sendcmdto_one(&me, CMD_NOTICE, sptr, "%C :If you are sure this is what you want, use /quote RESTART %s", sptr, cli_name(&me));
+    return 0;
+  }
+
   log_write(LS_SYSTEM, L_NOTICE, 0, "Server RESTART by %#C", sptr);
   server_restart("received RESTART");
 
