Fix a potential invalid memory access in iptobase64.

diff -r 66b953f37f4a ircd/numnicks.c
--- a/ircd/numnicks.c	Sat Feb 15 15:47:58 2014 +0100
+++ b/ircd/numnicks.c	Sun Feb 16 14:38:59 2014 +0100
@@ -458,7 +458,7 @@
 
     assert(count >= 25);
     /* Can start by printing out the leading non-zero parts. */
-    for (ii = 0; (addr->in6_16[ii]) && (ii < 8); ++ii) {
+    for (ii = 0; (ii < 8) && (addr->in6_16[ii]); ++ii) {
       inttobase64(output, ntohs(addr->in6_16[ii]), 3);
       output += 3;
     }
