show nick[user@host] in kills for remote users

diff -r 8273a0f68e66 ircd/m_kill.c
--- a/ircd/m_kill.c	Thu Mar 18 12:44:52 2010 +0100
+++ b/ircd/m_kill.c	Thu Mar 18 13:23:51 2010 +0100
@@ -125,10 +125,17 @@
    * Note: "victim->name" is used instead of "user" because we may
    *       have changed the target because of the nickname change.
    */
-  sendto_opmask_butone(0, snomask,
+  if (MyUser(victim))
+    sendto_opmask_butone(0, snomask,
                        "Received KILL message for %s from %s Path: %s!%s %s",
                        get_client_name(victim, SHOW_IP), get_client_name_and_opername(sptr),
                        inpath, path, msg);
+  else
+    sendto_opmask_butone(0, snomask,
+                       "Received KILL message for %s[%s@%s] from %s Path: %s!%s %s",
+                       cli_name(victim), cli_username(victim), ircd_ntoa(&cli_ip(victim)),
+                       get_client_name_and_opername(sptr),
+                       inpath, path, msg);
   log_write_kill(victim, sptr, inpath, path, msg);
 
   /*
