dont sent local kills by local irc operator upstream

diff -r a58c741e7e71 ircd/m_kill.c
--- a/ircd/m_kill.c
+++ b/ircd/m_kill.c
@@ -143,7 +143,9 @@
    *  - wiebe            
    */
    /* if (IsServer(cptr) || !MyConnect(victim)) { */
-  sendcmdto_serv_butone(sptr, CMD_KILL, cptr, "%C :%s!%s %s", victim,
+  /* dont sent kills by local opers upstream though */
+  if (!(MyUser(sptr) && IsLocOp(sptr) && MyUser(victim))) {
+     sendcmdto_serv_butone(sptr, CMD_KILL, cptr, "%C :%s!%s %s", victim,
                           inpath, path, msg);
 
    /*
@@ -151,7 +153,8 @@
     * the unnecessary QUIT for this. (This flag should never be
     * set in any other place)
     */
-  SetFlag(victim, FLAG_KILLED);
+    SetFlag(victim, FLAG_KILLED);
+  }
 
   /*
    * Tell the victim she/he has been zapped, but *only* if
