# HG changeset patch
# Parent 705e3be1f1345ea73bc25af87053b37780e6c2fd

diff -r 705e3be1f134 doc/quakenet.conf
--- a/doc/quakenet.conf
+++ b/doc/quakenet.conf
@@ -6,16 +6,40 @@
 #  v.12 config file, last edited 08.12.2007/paul
 # -----------------------------------------------------------------------------
 
+# -----------------------------------------------------------------------------
+# General
+# -----------------------------------------------------------------------------
+# name = "servername";
+# description = "server description"
+# numeric = 0-4095; must be unique on network
+#
+# IP of local interface to use for outgoing connections (optional)
+# vhost = "ipv4vhost";
+# vhost = "ipv6vhost";
+#
+# Local IP to use for DNS lookups (default let OS assign address)
+# dns vhost = "ipv4vhost";
+# dns vhost = "ipv6vhost";
+#
+# DNS server to use (default read from /etc/resolve.conf)
+# dns server = "ip1";
+# dns server = "ip2";
+#
 General {
   name = "@@@.@@@.quakenet.org";
   description = "@@@";
   numeric = @@@;
-  vhost = "@@@";
-  dns vhost  = "@@@";
-  dns server = "@@@";
-  dns server = "@@@";
+  #vhost = "@@@";
+  #dns vhost  = "@@@";
+  #dns server = "@@@";
+  #dns server = "@@@";
 };
 
+# -----------------------------------------------------------------------------
+# Admin
+# -----------------------------------------------------------------------------
+# This sets information that can be retrieved with the /ADMIN command.
+# It should contain at least an admin Email contact address.
 Admin {
   Location = "@@@";
   Location = "QuakeNet IRC server";
@@ -25,68 +49,99 @@
 # -----------------------------------------------------------------------------
 # Classes (aka Ylines)
 # -----------------------------------------------------------------------------
-# name     = "<class>";      pingfreq = time;  connectfreq = time;
-# maxlinks = number 0 or 1;  sendq    = size;     usermode = "+i";
+# name     = "<class>";  pingfreq = time;  connectfreq = time;
+# maxlinks = limit;      sendq    = size;     usermode = "+i";
 #
-# Uplinks
+# maxlinks
+#  for a user class:
+#    the number of connections allowed to the server in the class
+#  for a server class:
+#    keep auto connecting until the server is connected to this many servers in the class
+#    only considers servers with autoconnect set to yes
+#    when set to 0 no auto connects are done
+#
+# Leaf: uplink to hub server
 Class {
- name = "Server";
- pingfreq = 3 minutes;
- connectfreq = 5 minutes;
- maxlinks = 1;
- sendq = 40000000;
+  name = "Server";
+  pingfreq = 3 minutes;
+  connectfreq = 5 minutes;
+  # auto connect to uplink
+  maxlinks = 1;
+  sendq = 40000000;
 };
 
-# All the leaf servers you hub for.
+# Hub: all the leaf servers you hub for.
 Class {
- name = "LeafServer";
- pingfreq = 3 minutes;
- connectfreq = 5 minutes;
- maxlinks = 0;
- sendq = 40000000;
+  name = "LeafServer";
+  pingfreq = 3 minutes;
+  connectfreq = 5 minutes;
+  maxlinks = 0;
+  sendq = 40000000;
 };
 
 # Client class (i.e. everyone)
 Class {
- name = "Users";
- pingfreq = 3 minutes;
- sendq = 300000;
- maxlinks = 30000;
- usermode = "+i";
+  name = "Users";
+  pingfreq = 1 minutes 30 seconds;
+  sendq = 300000;
+  # the number of users allowed on the server
+  # this should not be greater than the number
+  # of connections the server is compiled for  
+  maxlinks = 30000;
+  usermode = "+iw";
 };
 
 # VIP user class (aka locals, opers etc..)
 Class {
- name = "Local";
- pingfreq = 3 minutes;
- sendq = 600000;
- maxlinks = 100;
- usermode = "+iw";
+  name = "Local";
+  pingfreq = 1 minutes 30 seconds;
+  sendq = 600000;
+  maxlinks = 100;
+  usermode = "+iw";
 };
 
 # Oper class (global)
 Class {
- name = "Opers";
- pingfreq = 3 minutes;
- sendq = 10240000;
- maxlinks = 10;
- local = no;
- # privs
- die = no;
- restart = no;
- rehash = no;
- jupe = no;
- local_jupe = no;
- freeform = yes; # change this if you want to disable freeform sethost for your opers 
+  name = "Opers";
+  pingfreq = 1 minutes 30 seconds;
+  sendq = 10240000;
+  maxlinks = 100;
+  local = no;
+  # change this if you want to disable freeform sethost for your opers
+  freeform = yes;
+  # privs for admins
+  die = no;
+  restart = no;
+  rehash = no;
+  jupe = no;
+  local_jupe = no;
+  # privs related to local channels
+  mode_lchan = no;
+  deop_lchan = no;
+  local_opmode = no;
+  force_local_opmode = no;
+  # channel quarantines not used 
+  force_opmode = no;
+  # whox usage not logged
+  whox = no;
 };
 
 # Oper class (global)
 Class {
- name = "Admin";
- pingfreq = 3 minutes;
- sendq = 10240000;
- maxlinks = 10;
- local = no;
+  name = "Admin";
+  pingfreq = 1 minutes 30 seconds;
+  sendq = 10240000;
+  maxlinks = 50;
+  local = no;
+  # privs related to local channels
+  mode_lchan = no;
+  deop_lchan = no;
+  local_opmode = no;
+  force_local_opmode = no;
+  # channel quarantines not used 
+  force_opmode = no;
+  # whox usage not logged
+  whox = no;
 };
 
 # -----------------------------------------------------------------------------
@@ -95,18 +150,17 @@
 # username = "ident";       host = "host";              ip = "127.0.0.0/8";
 # password = "password";   class = "classname";   maxlinks = 3;
 
-# Fall back entries (we let everyone in)
-Client { class = "Users"; ip = "*@*"; };
-Client { class = "Users"; host = "*@*"; };
+# Fall back entries (we let everyone in) - just IPv4 for now
+Client { class = "Users"; ip = "*@0.0.0.0/0"; };
 
 # Providers that we've decided to restrict due to abuse mainly.
-Client { class = "Users"; maxlinks = 2; host = "*@*.t-dialin.net"; };
-Client { class = "Users"; maxlinks = 1; host = "*@*.proxad.net"; };
-Client { class = "Users"; maxlinks = 1; host = "*@*.bezeqint.net"; };
-Client { class = "Users"; maxlinks = 1; host = "*@*.ono.com"; };
-Client { class = "Users"; maxlinks = 1; host = "*@*.wanadoo.nl"; };
-Client { class = "Users"; maxlinks = 2; host = "*@*.wanadoo.fr"; };
-Client { class = "Users"; maxlinks = 1; host = "*@*.rima-tde.net"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.dip0.t-ipconnect.de"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.proxad.net"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.bezeqint.net"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.ono.com"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.wanadoo.nl"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.wanadoo.fr"; };
+Client { class = "Users"; maxlinks = 3; ip = "*@0.0.0.0/0"; host = "*@*.rima-tde.net"; };
 
 # -----------------------------------------------------------------------------
 # UWorld/Jupe (aka ulines)
@@ -116,13 +170,15 @@
 UWorld {
   name = "CServe.quakenet.org";
   name = "operserv.quakenet.org";
-  name = "operserv2.quakenet.org";
+  name = "operserv?.quakenet.org";
   name = "proxyscan.quakenet.org";
-  name = "services4.uk.quakenet.org";
-  name = "services.no.quakenet.org";
-  name = "services.de.quakenet.org";
-  name = "services.se.quakenet.org";
-  name = "services.uk.quakenet.org";
+  name = "proxyscan?.quakenet.org";
+  name = "services.??.quakenet.org";
+  name = "services?.??.quakenet.org";
+  name = "services??.??.quakenet.org";
+  name = "services.??.??.quakenet.org";
+  name = "services?.??.??.quakenet.org";
+  name = "services??.??.??.quakenet.org";
 };
 
 Jupe {
@@ -157,38 +213,38 @@
 #
 
 Connect {
- name = "hub.@@@.quakenet.org";
- host = "@@@.@@@.@@@.@@@";
- password = "@@@";
- port = 4400;
- class = "Server";
- hub;
- autoconnect = yes;
+  name = "hub.@@@.quakenet.org";
+  host = "@@@.@@@.@@@.@@@";
+  password = "@@@";
+  port = 4400;
+  class = "Server";
+  hub;
+  autoconnect = yes;
 };
 
 # Our secondary uplink.
 #
 Connect {
- name = "hub.@@@.quakenet.org";
- host = "@@@.@@@.@@@.@@@";
- password = "@@@";
- port = 4400;
- class = "Server";
- hub;
- autoconnect = yes;
+  name = "hub.@@@.quakenet.org";
+  host = "@@@.@@@.@@@.@@@";
+  password = "@@@";
+  port = 4400;
+  class = "Server";
+  hub;
+  autoconnect = yes;
 };
 
 # Our tertiary uplink.
 #
 #
 Connect {
- name = "hub.@@@.quakenet.org";
- host = "@@@.@@@.@@@.@@@";
- password = "@@@";
- port = 4400;
- class = "Server";
- hub;
- autoconnect = yes;
+  name = "hub.@@@.quakenet.org";
+  host = "@@@.@@@.@@@.@@@";
+  password = "@@@";
+  port = 4400;
+  class = "Server";
+  hub;
+  autoconnect = yes;
 };
 
 
@@ -260,29 +316,34 @@
 # -----------------------------------------------------------------------------
 # Port (aka plines)
 # -----------------------------------------------------------------------------
-# port   = number;      mask   = "ipmask"    vhost = "virtualhostinterfaceip"
-# server = yes;         hidden = yes
+# port   = [ipv4|ipv6] number;
+# mask   = "ipmask"    vhost = "virtualhostinterfaceip"
+# server = yes         hidden = yes
+#
+# default is to bind to ALL interfaces, IPv4 and IPv6!
+# explicitly specify the type of interfaces to bind to!
 # mask only contains IP addresses or '*'
-# hidden, hides a port from stats
+# hidden, hides a port from stats from ordinary users (HIS already removes STATS p completely from them)
 
 Port {
-  port   = 4400;
+  port   = ipv4 4400;
   server = yes;
   hidden = yes;
   vhost  = "@@@";
 };
 
 Port {
-  vhost = "@@@" 6660;
-  vhost = "@@@" 6661;
-  vhost = "@@@" 6662;
-  vhost = "@@@" 6663;
-  vhost = "@@@" 6664;
-  vhost = "@@@" 6665;
-  vhost = "@@@" 6666;
-  vhost = "@@@" 6667;
-  vhost = "@@@" 6668;
-  vhost = "@@@" 6669;
+  port = ipv4 6660;
+  port = ipv4 6661;
+  port = ipv4 6662;
+  port = ipv4 6663;
+  port = ipv4 6664;
+  port = ipv4 6665;
+  port = ipv4 6666;
+  port = ipv4 6667;
+  port = ipv4 6668;
+  port = ipv4 6669;
+  port = ipv4 7000;
 };
 
 # -----------------------------------------------------------------------------
@@ -292,66 +353,34 @@
 
 features {
   "LOG" = "SYSTEM" "FILE" "ircd.log";
+  "LOG" = "SYSTEM" "LEVEL" "CRIT";
   "LOG" = "CONFIG" "FILE" "conf.log";
-  "LOG" = "SYSTEM" "LEVEL" "CRIT";
-  "DOMAINNAME"="quakenet.org";
-  "RELIABLE_CLOCK"="FALSE";
+  "LOG" = "OPER" "FILE" "oper.log";
+# "DOMAINNAME"="<obtained from /etc/resolv.conf by ./configure>";
   "BUFFERPOOL"="96000000";
-# "HAS_FERGUSON_FLUSHER"="FALSE";
-# "CLIENT_FLOOD"="1024";
-# "SERVER_PORT"="4400";
   "NODEFAULTMOTD"="FALSE";
-# "MOTD_BANNER"="TRUE";
-# "KILL_IPMISMATCH"="FALSE";
-# "IDLE_FROM_MSG"="TRUE";
-  "HUB"="FALSE";
-# "WALLOPS_OPER_ONLY"="FALSE";
-# "NODNS"="FALSE";
+# "MOTD_BANNER"="one-line banner to be sent if NODEFAULTMOTD is TRUE";
+# "HUB"="FALSE";
   "RANDOM_SEED"="@@@@@@@@@@@@@@@@";
-# "DEFAULT_LIST_PARAM"="TRUE";
-# "NICKNAMEHISTORYLENGTH"="800";
+# "DEFAULT_LIST_PARAM"="";
+  "NICKNAMEHISTORYLENGTH"="10000";
   "NETWORK"="QuakeNet";
-  "HOST_HIDING"="TRUE";
   "SETHOST" = "TRUE";
   "SETHOST_USER" = "TRUE";
   "SETHOST_AUTO" = "TRUE";
   "HIDDEN_HOST"="users.quakenet.org";
-  "HIDDEN_IP"="127.0.0.1";
   "URLREG"="http://www.quakenet.org/getauthed";
-# "KILLCHASETIMELIMIT"="30";
-  "MAXCHANNELSPERUSER"="20";
-  "NICKLEN" = "15";
-# "AVBANLEN"="40";
-# "MAXBANS"="45";
-# "MAXSILES"="15";
-# "HANGONGOODLINK"="300";
-# "HANGONRETRYDELAY" = "10";
-# "CONNECTTIMEOUT" = "90";
-# "MAXIMUM_LINKS" = "1";
-# "PINGFREQUENCY" = "120";
-# "CONNECTFREQUENCY" = "600";
-# "DEFAULTMAXSENDQLENGTH" = "40000";
-# "GLINEMAXUSERCOUNT" = "20";
-# "MPATH" = "ircd.motd";
-# "RPATH" = "remote.motd";
-# "PPATH" = "ircd.pid";
-# "TOS_SERVER" = "0x08";
-# "TOS_CLIENT" = "0x08";
-# "POLLS_PER_LOOP" = "200";
-# "IRCD_RES_TIMEOUT" = "4";
-# "IRCD_RES_RETRIES" = "2";
-# "AUTH_TIMEOUT" = "9";
-# "IPCHECK_CLONE_LIMIT" = "4";
-# "IPCHECK_CLONE_PERIOD" = "40";
-# "IPCHECK_CLONE_DELAY" = "600";
-# "CHANNELLEN" = "200";
+  "MAXCHANNELSPERUSER"="30";
+  "MAXBANS"="50";
+  "MAXSILES"="25";
   "CONFIG_OPERCMDS" = "TRUE";
   "OPLEVELS" = "FALSE";
   "ZANNELS" = "FALSE";
-# "LOCAL_CHANNELS" = "TRUE";
-# "ANNOUNCE_INVITES" = "FALSE";
+  "LOCAL_CHANNELS" = "FALSE";
+  "TOPIC_BURST" = "TRUE";
+  "ANNOUNCE_INVITES" = "TRUE";
   "HIS_SERVERNAME" = "*.quakenet.org";
   "HIS_SERVERINFO" = "QuakeNet IRC Server";
-  "HIS_URLSERVERS" = "http://www.quakenet.org/serverlist.php";
+  "HIS_URLSERVERS" = "http://www.quakenet.org/servers";
 };
 
