removed some old code in set_hostmask in s_user.c

diff -r ad23b9b1c0b6 ircd/s_user.c
--- a/ircd/s_user.c	Tue Mar 23 22:12:51 2010 +0100
+++ b/ircd/s_user.c	Tue Mar 23 22:15:10 2010 +0100
@@ -1146,28 +1146,6 @@
     send_reply(cptr, RPL_HOSTHIDDEN, hiddenhost);
   }
 
-#if 0
-  /* Code copied from hide_hostmask().  This is the old (pre-delayedjoin) 
-   * version.  Switch this in if you're not using the delayed join patch. */
-  /*
-   * Go through all channels the client was on, rejoin him
-   * and set the modes, if any
-   */
-  for (chan = cli_user(cptr)->channel; chan; chan = chan->next_channel) {
-    if (IsZombie(chan))
-      continue;
-    sendcmdto_channel_butserv_butone(cptr, CMD_JOIN, chan->channel, cptr,
-      "%H", chan->channel);
-    if (IsChanOp(chan) && HasVoice(chan)) {
-      sendcmdto_channel_butserv_butone(&me, CMD_MODE, chan->channel, cptr,
-        "%H +ov %C %C", chan->channel, cptr, cptr);
-    } else if (IsChanOp(chan) || HasVoice(chan)) {
-      sendcmdto_channel_butserv_butone(&me, CMD_MODE, chan->channel, cptr,
-        "%H +%c %C", chan->channel, IsChanOp(chan) ? 'o' : 'v', cptr);
-    }
-  }
-#endif
-
   /*
    * Go through all channels the client was on, rejoin him
    * and set the modes, if any
